Import the necessary Libraries

In [30]:
import pandas as pd
import numpy as np

Exploring Kaggle Dataset

In [31]:
# load the training dataset
train = pd.read_parquet("Training.parquet")
train.shape

(7658, 89)

In [32]:
# load the test dataset
test = pd.read_parquet("Testing.parquet")
test.shape

(3772, 89)

Combine the train and test datasets

In [33]:
combined_data = pd.concat([train, test], axis=0, ignore_index=True)
combined_data.shape

(11430, 89)

#### check the columns of the combined dataframe

In [34]:
combined_data.columns

Index(['url', 'length_url', 'length_hostname', 'ip', 'nb_dots', 'nb_hyphens',
       'nb_at', 'nb_qm', 'nb_and', 'nb_or', 'nb_eq', 'nb_underscore',
       'nb_tilde', 'nb_percent', 'nb_slash', 'nb_star', 'nb_colon', 'nb_comma',
       'nb_semicolumn', 'nb_dollar', 'nb_space', 'nb_www', 'nb_com',
       'nb_dslash', 'http_in_path', 'https_token', 'ratio_digits_url',
       'ratio_digits_host', 'punycode', 'port', 'tld_in_path',
       'tld_in_subdomain', 'abnormal_subdomain', 'nb_subdomains',
       'prefix_suffix', 'random_domain', 'shortening_service',
       'path_extension', 'nb_redirection', 'nb_external_redirection',
       'length_words_raw', 'char_repeat', 'shortest_words_raw',
       'shortest_word_host', 'shortest_word_path', 'longest_words_raw',
       'longest_word_host', 'longest_word_path', 'avg_words_raw',
       'avg_word_host', 'avg_word_path', 'phish_hints', 'domain_in_brand',
       'brand_in_subdomain', 'brand_in_path', 'suspecious_tld',
       'statistical_report', 

- The dataset has 89 different columns
- the status column is the target column

#### create a new dataframe that has only the url and status columns

In [35]:
# Create a new DataFrame with only the 'url' and 'status' columns
url_status_df = combined_data[['url', 'status']]
url_status_df.head()

url      status
0  https://www.todayshomeowner.com/how-to-make-ho...  legitimate
1  http://thapthan.ac.th/information/confirmation...    phishing
2  http://app.dialoginsight.com/T/OFC4/L2S/3888/B...    phishing
3                           https://www.bedslide.com  legitimate
4  https://tabs.ultimate-guitar.com/s/sex_pistols...  legitimate

#### convert status column to numerical

In [36]:
# Create a copy to avoid the SettingWithCopyWarning
url_status_df = url_status_df.copy()

# Map the 'status' column to numerical values
url_status_df['status'] = url_status_df['status'].map({'legitimate': 0, 'phishing': 1})
url_status_df.head()


url  status
0  https://www.todayshomeowner.com/how-to-make-ho...       0
1  http://thapthan.ac.th/information/confirmation...       1
2  http://app.dialoginsight.com/T/OFC4/L2S/3888/B...       1
3                           https://www.bedslide.com       0
4  https://tabs.ultimate-guitar.com/s/sex_pistols...       0

#### inspect the unique values in the status column 

In [37]:
url_status_df['status'].unique()

array([0, 1], dtype=int64)

#### Check the value counts for both the legitimate and the legitimate and the phishing urls

In [38]:
url_status_df.status.value_counts()

1    5715
0    5715
Name: status, dtype: int64

- The dataset is balanced.
- It has 5715 legitimate urls and 5715 phishing urls

## Exploring the UCI Irvine Dataset

#### Loading the dataset

In [42]:
dataset2= pd.read_csv("PhiUSIIL_Phishing_URL_Dataset.csv")
dataset2.head()

FILENAME                                 URL  URLLength  \
0  521848.txt    https://www.southbankmosaics.com         31   
1   31372.txt            https://www.uni-mainz.de         23   
2  597387.txt      https://www.voicefmradio.co.uk         29   
3  554095.txt         https://www.sfnmjournal.com         26   
4  151578.txt  https://www.rewildingargentina.org         33   

                       Domain  DomainLength  IsDomainIP  TLD  \
0    www.southbankmosaics.com            24           0  com   
1            www.uni-mainz.de            16           0   de   
2      www.voicefmradio.co.uk            22           0   uk   
3         www.sfnmjournal.com            19           0  com   
4  www.rewildingargentina.org            26           0  org   

   URLSimilarityIndex  CharContinuationRate  TLDLegitimateProb  ...  Pay  \
0               100.0              1.000000           0.522907  ...    0   
1               100.0              0.666667           0.032650  ...    0   
2               100.0              0.866667           0.028555  ...    0   
3               100.0              1.000000           0.522907  ...    1   
4               100.0              1.000000           0.079963  ...    1   

   Crypto  HasCopyrightInfo  NoOfImage  NoOfCSS  NoOfJS  NoOfSelfRef  \
0       0                 1         34       20      28          119   
1       0                 1         50        9       8           39   
2       0                 1         10        2       7           42   
3       1                 1          3       27      15           22   
4       0                 1        244       15      34           72   

   NoOfEmptyRef  NoOfExternalRef  label  
0             0              124      1  
1             0              217      1  
2             2                5      1  
3             1               31      1  
4             1               85      1  

[5 rows x 56 columns]

#### 

#### check the columns of dataset 2

In [40]:
dataset2.columns

Index(['FILENAME', 'URL', 'URLLength', 'Domain', 'DomainLength', 'IsDomainIP',
       'TLD', 'URLSimilarityIndex', 'CharContinuationRate',
       'TLDLegitimateProb', 'URLCharProb', 'TLDLength', 'NoOfSubDomain',
       'HasObfuscation', 'NoOfObfuscatedChar', 'ObfuscationRatio',
       'NoOfLettersInURL', 'LetterRatioInURL', 'NoOfDegitsInURL',
       'DegitRatioInURL', 'NoOfEqualsInURL', 'NoOfQMarkInURL',
       'NoOfAmpersandInURL', 'NoOfOtherSpecialCharsInURL',
       'SpacialCharRatioInURL', 'IsHTTPS', 'LineOfCode', 'LargestLineLength',
       'HasTitle', 'Title', 'DomainTitleMatchScore', 'URLTitleMatchScore',
       'HasFavicon', 'Robots', 'IsResponsive', 'NoOfURLRedirect',
       'NoOfSelfRedirect', 'HasDescription', 'NoOfPopup', 'NoOfiFrame',
       'HasExternalFormSubmit', 'HasSocialNet', 'HasSubmitButton',
       'HasHiddenFields', 'HasPasswordField', 'Bank', 'Pay', 'Crypto',
       'HasCopyrightInfo', 'NoOfImage', 'NoOfCSS', 'NoOfJS', 'NoOfSelfRef',
       'NoOfEmptyRef', 'NoOf

#### create a new dataframe that has only the URL and the target variable

In [43]:
# Create a new DataFrame with only the 'url' and 'status' columns
url_label_df = dataset2[['URL', 'label']]
url_label_df.head()

URL  label
0    https://www.southbankmosaics.com      1
1            https://www.uni-mainz.de      1
2      https://www.voicefmradio.co.uk      1
3         https://www.sfnmjournal.com      1
4  https://www.rewildingargentina.org      1

#### check the shape of the url_label_df

In [44]:
url_label_df.shape

(235795, 2)

- The dataset has 235,795 observations with the target column to determine whether the url is a legitimate or phishing url

### Merging both datasets

Here the kaggle dataset and the UCI Irvine dataset are marged before extracting the features for each dataset.

The datasets will be merged as follows:
1. Standardize Column Names: Rename similar columns in each dataset to a common name.
2. Concatenate: Use pd.concat to combine the datasets.